#Week 2

##Session 4

This is part 4 of the Python for Data Science Bootcamp. In this session we will explore two important libraries in Data Science; Pandas and Numpy.

###Outline

- Recap Session 3
- Going through OOP challenges from students
- Data Analysis with Pandas
- Exploring NumPy 
- Challenges
​


## Modules
In programming, a module is a piece of software that has a specific functionality. For example, when building a ping pong game, one module would be responsible for the game logic, and another module would be responsible for drawing the game on the screen. Each module is a different file, which can be edited separately. Consider a module to be the same as a code script, a file containing a set of functions you want to include in your application.
- A module is a Python object with arbitrarily named attributes that you can bind and reference. In other words, a module is a file consisting of Python code. A module can define functions, classes and variables. A module can also include runnable code.

Simply put

- ** Module. ** a third party script of code that has functions that you want to use in your script
- **Package.** collection of modules 
- **Library.** collection of packages

##Importing Libraries
###What's an import?
To use any package (package = a collection of modules) in your code, you must first make it accessible. You have to import it. You can't use anything in Python before it is defined. 
Simlpy put: from now on, we will import libraries (collection of packages) to use their functionalities in data science. You will get more familiar with this vague concept by doing. So let's start coding!


![alt_text](http://images7.memedroid.com/images/UPLOADED911/595fa17b69853.jpeg)

### Importing Pandas

During this session, we will mostly use Pandas to manipulate our datasets to our preferences
Pandas is an open-source Python library that provides data analysis tools. Pandas aims to provide a lot of the data manipulation and analysis functionality that people use R for. 'Pandas' is derived from the term 'Panel Data' that you are probably familiar with from statistics courses in your studies. If not, in a nutshell, Panel data is data that is derived from a (usually small) number of observations over time on a (usually large) number of cross-sectional units like individuals, households, firms, or governments. 

![alt_text](https://cdn-images-1.medium.com/max/1600/1*93CVLqnQESmvfOhzvYUgQw.png)

We will start of by importing the Pandas library and assigning it a shorter name. As we will be too lazy to fully write 'pandas' in our code when we will call the library in the future. So in the cell below we will import the pandas library and assign it the name 'pd'

The pandas module, just as any library you import, is imported at the top of your script

In [0]:
import pandas as pd

[This](https://www.youtube.com/watch?v=XDAnFZqJDvI) introduction video (5 min) will tell you more about the most popular methods within the pandas library 

What problem does pandas solve?
Python has long been great for data munging and preparation, but less so for data analysis and modeling. pandas helps fill this gap, enabling you to carry out your entire data analysis workflow in Python.


pandas does not implement significant modeling functionality outside of linear and panel regression. More work is still needed to make Python a first class statistical modeling environment, but we are well on our way toward that goal.

### Dataframes
A Dataframe is an object that stores data as rows and columns. You can think of a Dataframe as a spreadsheet. You can manually create a Dataframe or fill it with data from a CSV or an Excel spreadsheet.

Dataframes have rows and columns. Each column has a name, which is a string. Each row has an index (usually shown as a number in the most left column), which is an integer. Dataframes can contain many different data types: strings, ints, floats, tuples, etc.

Below is an image of a simple example of a Dataframe in Pandas (notice how similar it is to an Excel spreadsheet)

![alt_text](http://queirozf.com/images/contents/kHrnSlV.png)

##Uploading a csv file
Now that you are familiar with how DataFrames work, it is time for the next step. For this next step, we will load a .csv file using the read_csv function of the Pandas library


In [0]:
# First, we will import the pandas library and save it as pd
import pandas as pd

In [0]:
# We need to upload our csv file like this:
import pandas as pd
from google.colab import files
uploaded = files.upload()
import io

grades = pd.read_csv(io.StringIO(uploaded['grades.csv'].decode('utf-8')), delimiter=';')
# Dataset is now stored in a Pandas Dataframe

Saving bankloans.csv to bankloans.csv


NameError: ignored

In [0]:
# Next, we will print the DataFrame to show us how it looks
print(grades)

      name  gender  grade
0  Barbara  Female    7.5
1    Larry    Male    5.4
2     Gary    Male    6.7
3     Mary  Female    6.8
4    Brian    Male    9.7
5  Phillip    Male    8.7
6      Rex    Male    8.1
7   Lorina  Female    7.2
8   Corina  Female    9.1
9   Laurie  Female    8.7


In [0]:
# alright, now we only want to view the 'gender' column
# We will do this by selecting this column like this
print(grades.gender)

0    Female
1      Male
2      Male
3    Female
4      Male
5      Male
6      Male
7    Female
8    Female
9    Female
Name: gender, dtype: object


- If we only want to select a particular set of data in the DataFrame we can use the 'iloc[]' or 'loc[]' method
  - The iloc[] method is 'integer based'
  - The loc method is 'label based'
- Simply put, the iloc[] method takes in numbers and the loc[] method takes in names
- To illustrate this difference between iloc[] and loc[] we will show some examples:

For the iloc[] method, we use the index of the rows and columns as input for the method to determine what data we want to select. So the method is written as follows: name_of_dataframe.iloc[(row_index),(column index)]

In [0]:
# Here we only want to select the data related to the first student in our DataFrame
# Note that the rows are zero-indexed
# This means that the first student in our DataFrame has an index of 0
grades.iloc[0,:]

name      Barbara
gender     Female
grade         7.5
Name: 0, dtype: object

In [0]:
# We can also select mutiple rows using the iloc[] method
# Here we will select the rows starting at an index of 1 until the end of our DataFrame
grades.iloc[1:]

name  gender  grade
1    Larry    Male    5.4
2     Gary    Male    6.7
3     Mary  Female    6.8
4    Brian    Male    9.7
5  Phillip    Male    8.7
6      Rex    Male    8.1
7   Lorina  Female    7.2
8   Corina  Female    9.1
9   Laurie  Female    8.7

For the loc[] method, we use the name of the label of the rows and columns as input for the method to determine what data we want to select. So the method is written as follows: name_of_dataframe.loc[(name of label row),(name of label column)]
- For our gender dataframe, the label of the rows is the same as the index of the rows (as you can see to the very left when we print the dataframe). The label of the columns is equal to the names of our columns

In [0]:
# Here we select the value of the 'gender' columns for the row with an index of 3
grades.loc[3,'gender']

'Female'

We can also write a command that only selects the students that satisfy a statement
- For examlpe, we can only select the students with a median grade of above 8.0
-  Note that the logic behind this method is quite similar to an 'if' statement we used in week 1

In [0]:
grades[grades.grade>8.0]

name  gender  grade
4    Brian    Male    9.7
5  Phillip    Male    8.7
6      Rex    Male    8.1
8   Corina  Female    9.1
9   Laurie  Female    8.7

We can also add 'or' & 'and' expression to our method
- $|$ means 'or'
-  $&$ means 'and'
- So if we want to select the row of a student that has a grade above 8.0 and is named 'Laurie' we write:

In [0]:
grades[(grades.grade>8.0) & (grades.name=='Laurie')]

name  gender  grade
9  Laurie  Female    8.7

##In-class exercise
Select the rows in our DataFrame of the students that either got below a 6.0 or above a 9.0

Let's say all grades are corrected downwards by the professor. So the values of all new grades will only be subtracted with 0.2 from the old grades that the students got.
- Here's how we can easily create this new grade for every student:

In [0]:
grades['new_grade']=grades.grade-0.2

In [0]:
grades

name  gender  grade  new_grade
0  Barbara  Female    7.5        7.3
1    Larry    Male    5.4        5.2
2     Gary    Male    6.7        6.5
3     Mary  Female    6.8        6.6
4    Brian    Male    9.7        9.5
5  Phillip    Male    8.7        8.5
6      Rex    Male    8.1        7.9
7   Lorina  Female    7.2        7.0
8   Corina  Female    9.1        8.9
9   Laurie  Female    8.7        8.5

Now we want to get rid of the old grades in our DataFrame. 
- We can delete an entire column by using the 'drop' method
- We also have to select the axis that we want to drop
- The columns are on 'axis=1'

In [0]:
# So deleting the 'grade' can be written as follows:
grades = grades.drop('grade', axis=1)
print(grades)
# End of Pandas

      name  gender  new_grade
0  Barbara  Female        7.3
1    Larry    Male        5.2
2     Gary    Male        6.5
3     Mary  Female        6.6
4    Brian    Male        9.5
5  Phillip    Male        8.5
6      Rex    Male        7.9
7   Lorina  Female        7.0
8   Corina  Female        8.9
9   Laurie  Female        8.5


##Starbucks menu dataset
Now, we will perform operations on a dataset based on the nutritional values of the items on the Starbucks menu. I know what you are thinking; "how did they get this dataset?". The answer is simple, Starbucks posted it on Kaggle. 

![alt_text](https://www.kaggle.com/static/images/site-logo.png)

www.kaggle.com is an onine community of data scientists and machine learning engineers, owned by Google. Kaggle lets users publish and find data sets, explore and build models and also offers competitions for data scientists to solve with a chance of a monetary prize in most cases!


In [0]:
from google.colab import files
uploaded = files.upload()
import io
import pandas as pd

starbucks = pd.read_csv(io.StringIO(uploaded['starbucks.csv'].decode('utf-8')), delimiter=',')
# Dataset is now stored in a Pandas Dataframe

Saving starbucks.csv to starbucks.csv


In [0]:
print(starbucks)

                                            Unnamed: 0   Calories   Fat (g)  \
0                                         Chonga Bagel        300       5.0   
1                                         8-Grain Roll        380       6.0   
2                                     Almond Croissant        410      22.0   
3                                        Apple Fritter        460      23.0   
4                                     Banana Nut Bread        420      22.0   
5               Blueberry Muffin with Yogurt and Honey        380      16.0   
6                                      Blueberry Scone        420      17.0   
7                                     Butter Croissant        240      12.0   
8                                     Butterfly Cookie        350      22.0   
9                                        Cheese Danish        320      16.0   
10                              Chewy Chocolate Cookie        170       5.0   
11                               Chocolate Chip Cook

Because our Starbucks dataset is quite large, we do not want the computer to display the entire dataset
- So instead of using a "print" statement, we will use the .head() method
- This will only display the first five rows of our dataset

In [0]:
starbucks.head()

NameError: ignored

In [0]:
#IF you write head(n), it will display the first n number of rows. Like this:

starbucks.head(8)

NameError: ignored

In [0]:
# Becuase we are still getting familar wit our dataset, we could use a statistic summary of our data
# The .describe() method will display a broad statistical summary of our dataset
starbucks.describe()

Calories     Fat (g)   Carb. (g)   Fiber (g)   Protein (g)
count  113.000000  113.000000  113.000000  113.000000    113.000000
mean   356.637168   16.353982   41.486726    2.849558     11.469027
std    127.710685    8.297397   15.796764    2.888466      8.463230
min     90.000000    0.000000    5.000000    0.000000      1.000000
25%    280.000000    9.000000   31.000000    1.000000      5.000000
50%    360.000000   17.000000   42.000000    2.000000      8.000000
75%    450.000000   23.000000   53.000000    3.000000     19.000000
max    650.000000   37.000000   80.000000   21.000000     34.000000

In [0]:
# The first column header "Unnamed: 0" doesn't look nice
# We want to change the header of this column
# First, let's display all the names of the columns via the .columns method
starbucks.columns

Index(['Unnamed: 0', ' Calories', ' Fat (g)', ' Carb. (g)', ' Fiber (g)',
       ' Protein (g)'],
      dtype='object')

Ok, now we will use the `.rename` method to change the header of the column
- We can simpply copy + paste the header's name from the display of our previous command
- Then we will type the new header we want, 'Food' and finally we will write a command that will make sure the new title is put in place of the old one

In [0]:
starbucks.rename(columns = {'Unnamed: 0': 'Food'}, inplace=True)
starbucks.head()

Food   Calories   Fat (g)   Carb. (g)   Fiber (g)   Protein (g)
0      Chonga Bagel        300       5.0          50           3            12
1      8-Grain Roll        380       6.0          70           7            10
2  Almond Croissant        410      22.0          45           3            10
3     Apple Fritter        460      23.0          56           2             7
4  Banana Nut Bread        420      22.0          52           2             6

## NumPy

NumPy is a great library that can be used to store and manipulate numerical data in arrays. NumPy is the core library for scientific computing in Python. NumPy provides an abundance of useful features for operations on n-arrays and matrices in Python. Most (more advanced) Python packages that can be used for machine learning depend on features of the NumPy library to a reasonable extent. Because NumPy is very useful for performing mathematical and logical operations on Arrays. It provides an abundance of useful features for operations on n-arrays and matrices in Python.

![!alt_text](https://miro.medium.com/max/800/1*mc5YIn7jvo5uwuqBOUDw7Q.jpeg)

- If it still sounds a bit abstract, you may want to read [this](https://docs.scipy.org/doc/numpy-1.15.0/user/quickstart.html) article for beginners that clearly explains what some basic NumPy operations can do for you.
- If you are already familiar with MATLAB, you might find [this](https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html) tutorial useful to get started with NumPy. The tutorial gives a clear overview of the similarities and differences between NumPy and MATLAB.

##Arrays
###What is an array?
An array is a data structure that contains a group of elements. Typically these elements are all of the same data type, such as an integer or string. Arrays are commonly used in computer programs to organize data so that a related set of values can be easily sorted or searched.

In [0]:
# First we will import NumPy in a similar fashion as we did for Pandas
# We will call it 'np' for short in our code like this:
import numpy as np

In [0]:
# We will now create an array for test scores of students
# The array will be called 'test_1'
test_1 = np.array([7.2, 9.4, 8.8, 9.1, 8.7])
print(test_1)

[7.2 9.4 8.8 9.1 8.7]


In [0]:
# Now we will create a new array with the scores of the second test
# This array will be called 'test_2'
test_2 = np.array([7.9, 8.1, 8.6, 9.3, 9.1])

In [0]:
# Here we will print all our arrays to see what we have created
print(test_1)
print(test_2)

[7.2 9.4 8.8 9.1 8.7]
[7.9 8.1 8.6 9.3 9.1]


In [0]:
# Now we want to calculate the average grade of the students
# This can be done in the following way:
average_grade = (test_1 + test_2) / 2

In [0]:
print(average_grade)

[7.55 8.75 8.7  9.2  8.9 ]


## In class exercise
###The interpretation behind the arrays of our test scores is the following:
- There are three tests, taken by every student
- There are five students in our sample, that have each taken every test
- Exercise: print the test score of the first student in 'test_2'
![alt_text](https://i.imgur.com/VRSkSGd.jpg)

###We can also use logical operations for our NumPy arrays just like we did with 'if' statements and Pandas previously
We have an array of weather temperatures this past week below called 'temperature'
We want to find out for what days we can wear a summer jacket outside
We also want to know when we can wear a winter coat outside
We can do this by first creating our 'temperature array':

In [0]:
temperature = np.array([12,9,20,22,7,12,15])
winter_coat = temperature[(temperature<=10)]
summer_jacket = temperature[(temperature>=11) & (temperature<=19)]
print(winter_coat)
print(summer_jacket)
# The days with the following temperatures are the days when we wear a winter coat or summer jacket

[9 7]
[12 12 15]


##Matrices with NumPy
####Matrices are one of the core concepts of Machine Learning, that is why we will get our feet wet and start doing some matrix operations now. If you are not familiar with matrix calculations, or just need a refresher, [this](https://www.youtube.com/watch?v=sYlOjyPyX3g) video (10 min) will help you understand matrix calculations.

####Matrices and Machine learning
Linear algebra is the foundation for Machine Learning. Understanding the concepts behind matrices, vectors and other common operations is crucial for understanding how machine learning, specifically Deep learning works.

*The matrix below consists of 4 rows and 3 columns. We refer to this matrix as a 4 x 3 matrix*. 

![alt_text](http://machinethink.net/images/using-types-to-keep-yourself-honest/Matrix.png)

If you are wondering how NumPy can be useful for Machine Learning and Data Science, [this article](https://hackernoon.com/introduction-to-numpy-1-an-absolute-beginners-guide-to-machine-learning-and-data-science-5d87f13f0d51) may be very interesting to read.

In [0]:
# Now that you are familiar with an array in NumPy, it is time to create a matrix
# Here we will create matrix A by using NumPy
A = np.array([[1,5,3],
             [2,4,9],
             [6,7,8]])
print(A)

[[1 5 3]
 [2 4 9]
 [6 7 8]]


In [0]:
# Let's create another 3x3 matrix B with NumPy to get some more practice
B = np.array([[5,2,1],
             [3,6,4],
             [7,9,8]])
print(B)

[[5 2 1]
 [3 6 4]
 [7 9 8]]


In [0]:
# Matrices play a crucial role in the basics of Machine Learning
# Let's try some basic calculations with our matrices
C = A + B
print(C)

[[ 6  7  4]
 [ 5 10 13]
 [13 16 16]]


In [0]:
# We can multiply matrices (or any values we would like) by using the np.dot() method
D = np.dot(A,B)
print(D)

[[ 41  59  45]
 [ 85 109  90]
 [107 126  98]]


In [0]:
# As expected, the outcome of A*B is different than the outcome of B*A
E = np.dot(B,A)
print(E)

[[ 15  40  41]
 [ 39  67  95]
 [ 73 127 166]]


###Intuition: How to transpose a matrix
![alt_text](http://www.examiron.com/wp-content/uploads/2015/09/Screen-Shot-2015-09-17-at-5.51.20-PM.png)

In [0]:
# We can also transpose our matrices using the np.transpose() method
Atranspose = np.transpose(A)
print(Atranspose)

[[1 2 6]
 [5 4 7]
 [3 9 8]]


In [0]:
# Let's find the inverse of our matrix by using the 'inv' method of the linear algebra library within NumPy
Ainv = np.linalg.inv(A)
print(Ainv)

[[-0.24031008 -0.14728682  0.25581395]
 [ 0.29457364 -0.07751938 -0.02325581]
 [-0.07751938  0.17829457 -0.04651163]]


## Introduction to Challenge 'Temperature measurements'
We have created two 3x7 matrices in seperate csv files that have measured the temperature outside 3 times per day, 7 days a week. So for every day of the week, a temperature measurement was made in the morning, afternoon and evening. In total, we have observations of temperatures for two weeks because we have two datasets seperated over two csv files. If you still do not understand the context of our dataset, you will soon find out via printing the dataset later on.


In [0]:
import io
from google.colab import files
uploaded_temperatures = files.upload()

Saving temperatures.csv to temperatures (2).csv


In [0]:
#Now we will use the .read_csv function in Pandas to access the .csv file
temperatures = np.genfromtxt(io.StringIO(uploaded_temperatures['temperatures.csv'].decode('utf-8')), delimiter=';', dtype=int)

In [0]:
from google.colab import files
uploaded_temperatures2 = files.upload()

Saving temperatures2.csv to temperatures2.csv


In [0]:
temperatures2 = np.genfromtxt(io.StringIO(uploaded_temperatures2['temperatures2.csv'].decode('utf-8')), delimiter=';', dtype=int)

In [0]:
#Next, we will print the matrix to show us how it looks
print(temperatures)

[[-1  5  4 12 11 14  6]
 [ 4  3  6 16 19 21  7]
 [ 3  3  2 14 12 11  5]]


In [0]:
print(temperatures2)

[[-1  4  7 10 23 10 15]
 [ 6 17  6  7 28 23  7]
 [ 4 21  3  8 19 19  2]]


##Challenge 1
###Desciption of challenge:
- There has been a systematic measurement error in our dataset.
- It turns out that every measurement was 2 degrees too low.
- Fix the array so that every single measurement is increased by 2 degrees an save it under the name 'temperatures_fixed'
- print the 'temperatures_fixed' array

##Challenge 2
###Desciption of challenge:
- print the entire 'temperature_fixed' array for an overview
- Remember, every day there are three fixed measurement moments of the temeparture (morning, afternoon & evening (from top-to-bottom). The week runs from monday until sunday (left-to-right)
- save the value of the measured temperature on thursday afternoon under the name 'thursday_afternoon'
- print the value of 'thursday_afternoon'


##Challenge 3
###Desciption of challenge:
- We want to find out what the most extreme temperatures that we have measured are
- create a variable called 'temperatures_extremes'
- set 'temperatures_extremes' equal to the temperatures measured in the 'temperatures_fixed' that are either above 20 degrees or below 5 degrees
- print 'temperatures_extremes' to see which extreme temperatures we got

##Challenge 4
###Desciption of challenge:
- We want to print a 3 x 7 matrix that contains the average temperatures measured of the temperatures_fixed and temperatures2 observations
- Create a variable called 'temperatures_average' and print it


[[ 0.   5.5  6.5 12.  18.  13.  11.5]
 [ 6.  11.   7.  12.5 24.5 23.   8. ]
 [ 4.5 13.   3.5 12.  16.5 16.   4.5]]


##Challenge 5
###Desciption of challenge:
- Transpose the 'temperatures_average' matrix and save it as 'temperatures_average_transpose'
- Find all 3 average values of the temperature measurement on Wednesday- morining, afternoon and evening based on the location of these values in the 'temperatures_average_transpose' matrix
- Assign the values of the average temperatures measured on wednesday to the variable 'temperatures_average_wednesday'
- Print your 'temperatures_average_wednesday' variable. 
- Your answer should look like this: ''[6.5 7.  3.5]''

Hint: Remember to print your intermediate steps and variables to keep an overview of what you are doing while trying to reach your end goal

##Challenge 6
###Desciption of challenge:
- Add a row to our matrix 'temperatures_average_transpose'
  - This row should consist of three values of your choice



##Challenge 7
###Desciption of challenge:
- Create a new array
- The array should look like this:
- ([285, 272, 259, 246, 233, 220, 207, 194, 181, 168, 155, 142, 129,
       116, 103,  90,  77,  64,  51,  38,  25])
- Find a pattern and use only one line of code

###Resources:
- [NumPy explanation article](http://cs231n.github.io/python-numpy-tutorial/) 
- [Another article that explains some basic operations of NumPy](https://towardsdatascience.com/lets-talk-about-numpy-for-datascience-beginners-b8088722309f)
- [Recap of basic operations in Pandas](https://pandas.pydata.org/pandas-docs/stable/10min.html)
- [The Starbucks menu dataset on Kaggle](https://www.kaggle.com/starbucks/starbucks-menu/version/2)
- [10 Examples of Linear Algebra applications in Machine Learning](https://machinelearningmastery.com/examples-of-linear-algebra-in-machine-learning/)